PVB Linx Groep 6/Assets/Code/Scripts/Enemy/EnemyMovement.cs

```diff
@@ -0,0 +1,150 @@
+        [SerializeField] private float targetRange;
+        [SerializeField] [Range(0,5)] private float attackRange;
+
+        private Transform _currentTarget;
+        private bool _foundPlayer;
+        private bool _wait;
+        
+        // Start is called before the first frame update
+        private void Start()
+        {
+            getStates = EnemyStates.Target;
+            _currentTarget = mainStructure;
+        }
+
+        // Update is called once per frame
+        private void Update()
+        {
+            switch (getStates)
+            {
+                case EnemyStates.Target when !_foundPlayer:
+                    GetTarget(_currentTarget);
+                    break;
+                case EnemyStates.Move:
+                    MoveTowardsTarget();
+                    break;
+                case EnemyStates.Attack:
+                    AttackTarget();
+                    break;
+            }
+        }
+
+        private void GetTarget(Transform getTarget)
+        {
+            if (_foundPlayer)
+            {
+                _currentTarget = getTarget;
+                getStates = EnemyStates.Move;
+            }
+            else
+            {
+                var structures = GameObject.FindGameObjectsWithTag("Structure");
+                var shortestDis = Mathf.Infinity;
+                Transform nearestStructure = null;
+
+                foreach (var structureObj in structures)
+                {
+                    var structure = structureObj.transform;
+                    var distance = Vector3.Distance(transform.position, structure.position);
+                    if (distance < shortestDis)
+                    {
+                        shortestDis = distance;
+                        nearestStructure = structure;
+                    }
+                }
+
+                _currentTarget = nearestStructure;
+            }
+            
+            if (_currentTarget != null)
+                getStates = EnemyStates.Move;
+        }
+
+        private void MoveTowardsTarget()
+        {
+            if (_currentTarget != null)
+                enemyAgent.SetDestination(_currentTarget.position);
+
+            if (Vector3.Distance(transform.position, _currentTarget.position) <= attackRange)
+            {
+                getStates = EnemyStates.Attack;
+                AttackTarget();
+            }
+        }
+
+        private void AttackTarget()
+        {
+            StartCoroutine(Delay(4, 15));
+
+            if (_currentTarget == null)
+            {
+                getStates = EnemyStates.Target;
+                GetTarget(mainStructure);
+            }
+
+            if (Vector3.Distance(transform.position, _currentTarget.position) >= attackRange)
+            {
+                getStates = EnemyStates.Target;
+                GetTarget(_currentTarget);
+            }
+        }
+
+        private IEnumerator Delay(int delay, int damage)
+        {
```

Comment:
Having the actual attack happen in a function called delay is a little confusing, either move the actual attack to the AttackTarget function or rename this into something that makes it more clear what is actually happening.